grpc/interceptors.go

```diff
@@ -43,3 +44,62 @@ func EnsureTimeoutUnaryClientInterceptor(
 
 	return invoker(ctx, method, req, reply, cc, opts...)
 }
+
+// The following code is for appending/extracting grpc metadata regarding module names/origins via
+// contexts.
+type modNameKeyType int
+
+const modNameKeyID = modNameKeyType(iota)
+
+// GetModuleName returns the module name (if any) the request came from. The module name will match
+// a string from the robot config.
+func GetModuleName(ctx context.Context) string {
+	valI := ctx.Value(modNameKeyID)
+	if val, ok := valI.(string); ok {
+		return val
+	}
+
+	return ""
+}
+
+const modNameMetadataKey = "modName"
+
+// ModInterceptors takes a user input `ModName` and exposes an interceptor method that will attach
+// it to outgoing gRPC requests.
+type ModInterceptors struct {
+	ModName string
+}
+
+// UnaryClientInterceptor adds a module name to any outgoing unary gRPC request.
+func (mc *ModInterceptors) UnaryClientInterceptor(
+	ctx context.Context,
+	method string,
+	req, reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	ctx = metadata.AppendToOutgoingContext(ctx, modNameMetadataKey, mc.ModName)
+	return invoker(ctx, method, req, reply, cc, opts...)
+}
+
+// ModNameUnaryServerInterceptor checks the incoming RPC metadata for a module name and attaches any
+// information to a context that can be retrieved with `GetModuleName`.
+func ModNameUnaryServerInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	meta, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return handler(ctx, req)
+	}
+
+	values := meta.Get(modNameMetadataKey)
+	if len(values) == 1 {
+		ctx = context.WithValue(ctx, modNameKeyID, values[0])
+	}
+
+	return handler(ctx, req)
+}
```

Comment:
Do you not need `Stream` versions of the interceptors? Or do we only care about unary for now?

Reply:
I take this back -- streams don't have contexts. We [play games](https://github.com/viamrobotics/rdk/blob/36db2d2b6209b80692a1955fa4b02e1c8b5bb833/operation/web.go?plain=1#L87) to make that a thing. I'm inclined to punt for now.

Reply:
> I'm inclined to believe a stream would never need this?

Well we certainly don't have any gRPC streaming API between a module and rdk right now. Whether we never will: I'm not sure. I think punting is totally fine for now. I do remember looking at that `ServerStreamContextWrapper` thing and scratching my head... I guess we'd have to do that in the future if we ever need the same feature for streaming gRPC APIs module <-> rdk 🤷🏻 .